### PR TITLE
feat: derife constructor options from chainable `Base.defaults()` calls

### PR DESCRIPTION
**Pull request by gr2m**
Given that a `version` constructor parameter is required, the following code should work without type errors.

```js
const Base123 = Base.defaults({ version: "1.2.3" });
const base = new Base123();
```

### CODE REVIEW

**Comment by gr2m** [comment]
I asked a question about this on StackOverflow:
https://stackoverflow.com/questions/68201386/how-to-make-required-constructor-options-optional-when-they-were-set-using-mycla

**Comment by gr2m** [comment]
progress (I think): I managed to create a function which accepts both defaults and options, and all the type behavior behaves as expected:

```ts
type Options = {
  requiredOption: string;
  optionalOption?: string;
};

function testWithDefaults<
  TDefaults extends Partial<Options>,
  TOptions extends Optional<Options, keyof TDefaults> & Record<string, unknown>
>(defaults: TDefaults, options: TOptions) {
  return {
    ...defaults,
    ...options,
  };
}

const test = testWithDefaults({ requiredOption: ""}, {})
```

[playground link](https://www.typescriptlang.org/play?strictNullChecks=false#code/GYVwdgxgLglg9mABAUwB4AdnQDwBUB8AFAO4AWAhlAFyK4CUiA3gL4BQrUAnpogPLqwE5ADZ4UqKMjAATAM6I4AIwBWWKABpEAaXGSZ82VABOMMAHNEAH0RgQAW0XIjVxLM4O4w-IgC8fOzBQ2KyItOoh2qzeAGQRAArkRrAi2HEwEADWeJoZyJxwwLSI0dr4+ADc7Fw8-IJg8n6MEUbIAI4gMC3StfBgNIYm5pWhcAK9Ij0IAPz9xqZmlcyVrKCQdYiShoSjdbI0k-UMTaEtUCBGSDu9sovsq9C9G8iGAOqBpAAiyMDkIMJQsmCoVwXx+fwBuikckQCSSMBSB1k+HCwMRkP0fDGQlEiJyeQKtFBv3+SOKiAASlg4EZpNgBvNNOAMmA4MQwPgooRpN9iQCaCCeeDZJorgg9rREUdmsgzhcmBFQgA6ZXcsEklGhRDKxWi+oapasNisAD0xqehlYaEwOEYiBa7U6yG6WL6rjm5kQzCImyghFt9o6XQONAARPkQ566HRlqbzVA3lBPoKSZaMGpsP62oGncG3YMLF7CBEfQmk2qAUXNfKqycs47nXVQyGNVWWaGyJRmwrPS2mGxQnRWNHU9agpmHUGXbN856iMXnvH3kShZXNSxe8cawH67mQ12q-3EIPh1b0+Psw3etP5uVEKredePYX568l8mK93N1X7+Cm73axOOZToge69swG6HselSnjadp1pOjZ5jeCjAfST5zqEJZvuWsirqEX6atuCFXiB+6auBn7diMwGgd2kFDtBaawURQGIWhBYYXGpbLiSeHVlu8GsSRtEHhBERQSOZ5wYBl4II+CwoXUEyoe6Cm6vJs6rlhiY8R+VYEQBF67mRoQUfpVGKeMwjGf+llyaRYHiQxHALtx764RE64RLGAACAIALQwVAgVGEY1KIP50lGS6iAwPILJQHe3ymE6sVIImyB2fUnlsMOrAQGKiWrBkvhcdhvIefh3YsbJroiZqP7-H+ERmfx1GIfVuWVEAA)

**Comment by gr2m** [comment]
I've spent way too much time on this already.

The [answers to my the StackOverflow question](https://stackoverflow.com/questions/68201386/how-to-make-required-constructor-options-optional-when-they-were-set-using-mycla) do not work, because they both only implement a single `Base.defaults()` call and do not support chaining (`Base.defaults().defaults()`).

I've reached out to @joshuakgoldberg for more help and their assessment is that it's not possible with today's TypeScript features. Josh opened two issues on the TS repository

- https://github.com/microsoft/TypeScript/issues/45013
- https://github.com/microsoft/TypeScript/issues/45014

two other related issues are

- https://github.com/microsoft/TypeScript/issues/41581
- https://github.com/microsoft/TypeScript/issues/5863

I'd like to try one more approach: limit the amount of chaining to 3x and implement the carrying of state when chaining static class methods by duplicating & nesting code.

File: index.d.ts [hunk 1]
```diff
@@ -45,18 +80,55 @@ export declare class Base<TOptions extends Base.Options = Base.Options> {
...
+   * const base = new MyBase({ option: 'value' }); // `version` option is not required
+   * base.options // typed as `{ version: string, otherDefault: string, option: string }`
+   * ```
+   * @remarks
+   * Ideally, we would want to make this infinitely recursive: allowing any number of
+   * .defaults({ ... }).defaults({ ... }).defaults({ ... }).defaults({ ... })...
+   * However, we don't see a clean way in today's TypeScript syntax to do so.
+   * We instead artificially limit accurate type inference to just three levels,
+   * since real users are not likely to go past that.
+   * @see https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/pull/57
```
Comment (gr2m) [inline index.d.ts:98]: so what's interesting is that I tried adding a test with `// @ts-expect-error` for four levels of chaining, but  .... it just works? It seems to be working infinitely? 

```ts
const Test3 = Base.defaults({ version: "1.2.3", one: 1 })
  .defaults({ two: 2 })
  .defaults({ three: 3 });

const Test6 = Test3.defaults({ four: 4 })
  .defaults({ five: 5 })
  .defaults({ six: 6 })
  .plugin(withOptionsPlugin);

expectType<{
  version: string;
  one: number;
  two: number;
  three: number;
  four: number;
  five: number;
  six: number;
  // @ts-expect-error - Test3 does not have four / five / six properties
}>({ ...Test3.defaultOptions });

expectType<{
  version: string;
  one: number;
  two: number;
  three: number;
  four: number;
  five: number;
  six: number;
}>({ ...Test6.defaultOptions });
```

What is this sorcery?

Comment (gr2m) [reply index.d.ts:98]: Ah I got it now, what's limited is the `.options` property on the instances. I'll add a suggestion to the tests to cover that case

File: index.test-d.ts [hunk 2]
```diff
@@ -51,6 +148,9 @@ expectType<string>(baseWithVoidAndNonVoidPlugins.bar);
 baseWithVoidAndNonVoidPlugins.unknown;
 
 const BaseWithOptionsPlugin = Base.plugin(withOptionsPlugin);
-const baseWithOptionsPlugin = new BaseWithOptionsPlugin();
+const baseWithOptionsPlugin = new BaseWithOptionsPlugin({
+  version: "1.2.3",
+});
 
 expectType<string>(baseWithOptionsPlugin.getFooOption());
+
```
Comment (gr2m) [inline index.test-d.ts:156]: See my resolved discussion above: https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/pull/57/files#r670634320

```suggestion

// Test depth limits of `.defaults()` chaining
const BaseLevelFour = BaseLevelThree.defaults({ defaultFour: 4 });

expectType<{
  version: string;
  defaultOne: string;
  defaultTwo: number;
  defaultThree: string[];
  defaultFour: number;
}>({ ...BaseLevelFour.defaultOptions });

const baseLevelFour = new BaseLevelFour();

expectType<{
  version: string;
  defaultOne: string;
  defaultTwo: number;
  defaultThree: string[];
  defaultFour: number;
  // @ts-expect-error - .options from .defaults() is only supported until a depth of 4
}>({ ...baseLevelFour.options });

```

File: index.d.ts (hunk 1 above)
Comment (gr2m) [reply index.d.ts:98]: @JoshuaKGoldberg I suggest we clarify the current behavior in the README. 

TypeScript will not complain when chaining `.defaults()` calls endlessly and the static `.defaultOptions` property is set correctly. But when instantiating from a class with 4+ chained `.defaults()` calls, then only the defaults from the first 3 calls are supported. Is that correct?

File: index.test-d.ts [hunk 3]
```diff
@@ -6,30 +6,125 @@ import { barPlugin } from "./plugins/bar/index.js";
...
+const BaseLevelTwo = BaseLevelOne.defaults({
+  defaultTwo: 0,
+});
+
+expectType<{
+  defaultOne: string,
+  defaultTwo: number,
+  version: string,
+}>({ ...BaseLevelTwo.defaultOptions });
```
Comment (gr2m) [inline index.test-d.ts:60]: why the destructoring? Isn't this the same as 

```suggestion
}>(BaseLevelTwo.defaultOptions);
```

**Comment by gr2m** [comment]
I accidentally closed https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/pull/59 by pushing it to  the `defaults-options-inheritance` branch in this repository instead of @JoshuaKGoldberg's fork.

Josh wrote

> Builds onto #57 with the long-lived temporary solution @gr2m referenced in that PR thread of having a known finite level of nesting.
> 
> I did away with fancy conditional types to make sure the type system doesn't get bogged down computing them all for plugins with a few options. I can add them back if you have test cases that require more than just `version` being required?
> 
> The big return type of the `static defaults` in theory could be automated a little bit with a recursive type. However, we wouldn't be able to go infinitely deep, and again the code to get that to work would be a good bit longer and less readable than how it is now. 🤷

I think what broke is the combination of `.plugin().defaults()` call now

Here is a complex test I'd like to see pass

```ts
const BaseWithManyChainedDefaultsAndPlugins = Base.defaults({
  defaultOne: "value",
})
  .plugin(fooPlugin, barPlugin, voidPlugin)
  .defaults({
    defaultTwo: 0,
  })
  .plugin(withOptionsPlugin)
  .defaults({
    defaultThree: ["a", "b", "c"],
  });

expectType<{
  defaultOne: string;
  defaultTwo: number;
  defaultThree: string[];
}>({ ...BaseWithManyChainedDefaultsAndPlugins.defaultOptions });

const baseWithManyChainedDefaultsAndPlugins =
  new BaseWithManyChainedDefaultsAndPlugins({
    version: "1.2.3",
    foo: "bar",
  });

expectType<string>(baseWithManyChainedDefaultsAndPlugins.foo);
expectType<string>(baseWithManyChainedDefaultsAndPlugins.bar);
expectType<string>(baseWithManyChainedDefaultsAndPlugins.getFooOption());
```

Right now the last three `expectType` calls checking the mixins from the plugins fail

A minimal test case to reproduce this problem is

```ts
const BaseWithManyChainedDefaultsAndPlugins = Base.defaults({
  defaultOne: "value",
})
  .plugin(fooPlugin)
  .defaults({
    defaultTwo: 0,
  });

const baseWithManyChainedDefaultsAndPlugins =
  new BaseWithManyChainedDefaultsAndPlugins({
    version: "1.2.3",
  });

expectType<string>(baseWithManyChainedDefaultsAndPlugins.foo);
```

Calling `Base.defaults().plugin().defaults()` breaks the mixins from the plugins. Both `Base.defaults().plugin()` and `Base.plugin().defaults()` works

File: index.test-d.ts (hunk 3 above)
Comment (JoshuaKGoldberg) [reply index.test-d.ts:60]: Ha, this one was nifty. `tsd` sees the an intersection type as not being the same as its equivalent solo type.

```
  index.test-d.ts:86:0
  ✖  86:0  Parameter type { defaultOne: string; defaultTwo: number; defaultThree: string[]; version: string; } is not identical to argument type { defaultOne: string; version: string; } & { defaultTwo: number; } & { defaultThree: string[]; }.  
```

File: index.d.ts [hunk 4]
```diff
@@ -30,33 +31,103 @@ declare type ReturnTypeOf<T extends AnyFunction | AnyFunction[]> =
...
-    this: S,
-    plugin1: T1,
-    ...additionalPlugins: T2
-  ): S & {
-    plugins: any[];
-  } & Constructor<UnionToIntersection<ReturnTypeOf<T1> & ReturnTypeOf<T2>>>;
+    this: Class,
+    ...plugins: Plugins,
+  ): Class & {
+    plugins: [Class['plugins'], ...Plugins];
```
Comment (JoshuaKGoldberg) [inline index.d.ts:81]: I think this is right? At the very least it makes `plugins: Plugin[]`, which is more accurate than `any[]`, right?

File: index.d.ts [hunk 5]
```diff
@@ -30,33 +31,103 @@ declare type ReturnTypeOf<T extends AnyFunction | AnyFunction[]> =
...
+   * ```
+   */
   static plugin<
-    S extends Constructor<any> & {
-      plugins: any[];
-    },
-    T1 extends Plugin,
-    T2 extends Plugin[]
+    Class extends ClassWithPlugins,
+    Plugins extends [Plugin, ...Plugin[]],
```
Comment (JoshuaKGoldberg) [inline index.d.ts:76]: New little nifty improvement recently added to tuples in TypeScript ✨ no more multiple parameters!

File: index.test-d.ts [hunk 6]
```diff
@@ -51,6 +148,74 @@ expectType<string>(baseWithVoidAndNonVoidPlugins.bar);
...
+  .defaults({
+    defaultTwo: 0,
+  });
+
+const baseWithChainedDefaultsAndPlugins =
+  new BaseWithChainedDefaultsAndPlugins({
+    version: "1.2.3",
+  });
+
+expectType<string>(baseWithChainedDefaultsAndPlugins.foo);
```
Comment (JoshuaKGoldberg) [inline index.test-d.ts:193]: This is a tough one! It doesn't repro if either of the `.defaults({ ... })` are removed above... Tricky!

Comment (JoshuaKGoldberg) [reply index.test-d.ts:193]: Got it! It was the `this: Class` generic.

My leading hunch was that whatever information added the `foo: string` from the `.plugin(fooPlugin)` call was being abandoned by the `.defaults` calls going on top of each other. That meant the piece of the type that contained `foo` wasn't getting preserved.

That piece of type was the `this: Class` -> `& Class`. Using just one `<Class>` generic from the start meant new type information on new `this` scopes that had extra stuff added to them was being lost.

File: index.d.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply index.d.ts:98]: @gr2m I think that's accurate, yeah.

File: index.test-d.ts [hunk 7]
```diff
@@ -167,11 +167,11 @@ expectType<{
...
 
+// See the node on static defaults in index.d.ts for why defaultFour is missing
+// .options from .defaults() is only supported until a depth of 4
 expectType<{
   version: string;
   defaultOne: string;
   defaultTwo: number;
   defaultThree: string[];
-  defaultFour: number;
-  // @ts-expect-error - .options from .defaults() is only supported until a depth of 4
```
Comment (gr2m) [inline index.test-d.ts]: why did you remove it? Maybe we find a fix for this in future at which point this line will fail with `tsd` and remind us to update tests & docs?

File: index.d.ts (hunk 4 above)
Comment (gr2m) [reply index.d.ts:81]: Just from looking at it, wouldn't that result in a nested array? Shouldn't it be

```suggestion
    plugins: [...Class['plugins'], ...Plugins];
```

**Review by gr2m — COMMENTED** [review]
just a question, looks great otherwise! Super cool that you got it all working!

**Review by gr2m — COMMENTED** [review]
oops sorry I forgot to submit my review yesterday

File: index.test-d.ts (hunk 7 above)
Comment (JoshuaKGoldberg) [reply index.test-d.ts]: Ah, sorry, I'd meant to leave a note - `@ts-expect-error` could apply to the whole assertion. I'll duplicate to make sure we get both checks.

**Pull request merged**